### PR TITLE
Change CyFunction to compile with Py_LIMITED_API

### PR DESCRIPTION
**Pull request by da-woods**
Mostly that we store the PyCFunctionObject as an attribute of it rather than "inheriting" from PyCFunctionObject.

(Actually testing the compilation with `Py_LIMITED_API` defined needs to it be put on top of https://github.com/cython/cython/pull/5550, but I think this PR can stand alone generally)

### CODE REVIEW

**Comment by scoder** [comment]
Quick comment (this is great work overall), I don't think we discard the comment in

#else // CYTHON_COMPILING_IN_LIMITED_API

We should be using C89 here.

**Comment by scoder** [comment]
+#if !CYTHON_COMPILING_IN_LIMITED_API
     Py_CLEAR(((PyCFunctionObject*)m)->m_module);
+#else
+    Py_CLEAR(m->func);
+#endif

If there are alternative implementations, it's usually better to put the special case first, with a positive macro guard. If we add more special cases at some point, we'll end up with mixed positive and negative guards, and would probably reverse the condition at that point. So why not avoid that risk of future churn all together.

File: Cython/Utility/CythonFunction.c [hunk 1]
```diff
@@ -28,16 +28,22 @@
...
+    // We can't "inherit" from func, but we can use it as a data store
+    PyObject *func;
+#else
 #if PY_VERSION_HEX < 0x030900B1
     PyCFunctionObject func;
 #else
     // PEP-573: PyCFunctionObject + mm_class
     PyCMethodObject func;
 #endif
+#endif
```
Comment (scoder) [inline Cython/Utility/CythonFunction.c:42]:
```suggestion
#endif
```

File: Cython/Utility/CythonFunction.c [hunk 2]
```diff
@@ -164,11 +175,15 @@ __Pyx_CyFunction_get_name(__pyx_CyFunctionObject *op, void *context)
 {
     CYTHON_UNUSED_VAR(context);
     if (unlikely(op->func_name == NULL)) {
+#if CYTHON_COMPILING_IN_LIMITED_API
+        op->func_name = PyObject_GetAttrString(op->func, "__name__");
+#else
 #if PY_MAJOR_VERSION >= 3
```
Comment (scoder) [inline Cython/Utility/CythonFunction.c:181]:
```suggestion
#elif PY_MAJOR_VERSION >= 3
```

File: Cython/Utility/CythonFunction.c [hunk 3]
```diff
@@ -164,11 +175,15 @@ __Pyx_CyFunction_get_name(__pyx_CyFunctionObject *op, void *context)
...
         op->func_name = PyUnicode_InternFromString(((PyCFunctionObject*)op)->m_ml->ml_name);
 #else
         op->func_name = PyString_InternFromString(((PyCFunctionObject*)op)->m_ml->ml_name);
 #endif
+#endif  /* CYTHON_COMPILING_IN_LIMITED_API */
```
Comment (scoder) [inline Cython/Utility/CythonFunction.c:186]:
```suggestion
#endif  /* CYTHON_COMPILING_IN_LIMITED_API */
```

**Comment by da-woods** [comment]
There's a few bits of the destructor which I haven't got quite right... Don't merge yet!

**Comment by scoder** [comment]
Ok. Anyway, LGTM now. Even the ABI module change is fine, because it only touches modules that use the still officially experimental support for the Limited C-API.

**Comment by da-woods** [comment]
Got to the bottom of my referencing counting problems with this. I'll merge it as soon as the tests pass.

**Pull request merged**